### PR TITLE
Filter limit orders during auction cutting

### PR DESCRIPTION
**Pull request by sistemd**
Fixes https://github.com/cowprotocol/services/issues/740.

When cutting the auction, for each limit order convert the sell and buy price to the native token (ETH) and make sure that sell with the surplus fee is higher than buy with the configurable price factor. If it isn't, filter it out.

I think this approach makes sense but there might be some better idea that I'm missing.

### Test Plan

This module is structured as a bunch of freestanding functions that are being tested individually. I think it'd be more natural if these were methods, and in general I don't like this approach to testing, so I didn't want to keep adding to it. Related: https://github.com/cowprotocol/services/pull/713#discussion_r1014195804.

After https://github.com/cowprotocol/services/pull/713 is merged which should be soon, we can extend the E2E tests from https://github.com/cowprotocol/services/pull/746 to cover the full flow. And then we can start advertising and testing this feature in production :slightly_smiling_face:

### CODE REVIEW

File: crates/autopilot/src/solvable_orders.rs [hunk 1]
```diff
@@ -236,6 +249,29 @@ impl SolvableOrdersCache {
...
+    ) -> Vec<Order> {
+        orders.retain(|order| {
+            if order.metadata.class != OrderClass::Limit {
+                return true;
+            }
+
+            // Convert the sell and buy price to the native token (ETH) and make sure that sell
+            // with the surplus fee is higher than buy with the configurable price factor.
+            let sell_native = order.data.sell_amount * prices.get(&order.data.sell_token).unwrap()
+                + order.metadata.surplus_fee;
```
Comment (vkgnosis) [inline crates/autopilot/src/solvable_orders.rs:267]: Is it correct that surplus_fee is not multiplied by the price? Is the unit in the sell token or native token? Maybe we should comment that on to OrderMetadata::surplus_fee .

**Review by MartinquaXD — APPROVED** [review]
Code looks good, math looks reasonable.
One additional step we could consider is to have an upper limit on orders in an auction. Then we could take the n most favorably priced orders in case there are too many orders we consider reasonable.

I think this will not be necessary for a long time and should happen in another PR (if we even want that) but I just wanted to put that out there so we don't forget about it.

File: crates/autopilot/src/arguments.rs [hunk 2]
```diff
@@ -130,6 +130,9 @@ pub struct Arguments {
     /// in COW base units
     #[clap(long, env)]
     pub cip_14_reward_cap: Option<f64>,
+
+    #[clap(long, env, default_value = "0.97")]
```
Comment (nlordell) [inline crates/autopilot/src/arguments.rs:134]: I think 3% spread might be a little too tight for long-tail tokens. I think something more like 10% may be more reasonable. Maybe @harisang can comment as they have used some heuristics for filtering out mis-prised orders in our optimization solvers in the past.

That being said, I would actually use a default that allows all orders through (so `0.0`) and then only manually configure this value in our deployment configuration. My rationale is that when we are doing local testing, I don't want to have to think about what price my limit order needs to be to exercise the quoting code. Don't feel strongly about this point though.

File: crates/autopilot/src/solvable_orders.rs [hunk 3]
```diff
@@ -236,6 +249,29 @@ impl SolvableOrdersCache {
...
+            }
+
+            // Convert the sell and buy price to the native token (ETH) and make sure that sell
+            // with the surplus fee is higher than buy with the configurable price factor.
+            let sell_native = (order.data.sell_amount + order.metadata.surplus_fee)
+                * prices.get(&order.data.sell_token).unwrap();
+            let buy_native = order.data.buy_amount * prices.get(&order.data.buy_token).unwrap();
+            let sell_native = u256_to_big_decimal(&sell_native);
+            let buy_native = u256_to_big_decimal(&buy_native);
+            sell_native >= buy_native * self.limit_order_price_factor.clone()
```
Comment (nlordell) [inline crates/autopilot/src/solvable_orders.rs:271]: Nit: Sometimes its hard to debug why an order wasn't included in a batch. Can you add a log message including the order UID for when it gets filtered out?

**Review by nlordell — COMMENTED** [review]
Changes look good. Just some small nits.

**Comment by nlordell** [comment]
> One additional step we could consider is to have an upper limit on orders in an auction. Then we could take the n most favorably priced orders in case there are too many orders we consider reasonable.

I think this is another heuristic that we can add to better "choose" what orders to include in an auction. Agree it should be a separate PR though.

Once limit orders is released, we should do some testing on how much we can saturate the orderbook before we start running into performance issues. This will help us pick parameters like this one, as well the maximum number of orders per user.

**Review by vkgnosis — APPROVED** [review]

File: crates/autopilot/src/arguments.rs (hunk 2 above)
Comment (harisang) [reply crates/autopilot/src/arguments.rs:134]: I guess Marco and Tom may have done these nice preprocessing steps, but I am not really familiar with them.

However, 3% definitely sounds low. I would be very liberal with this number (e.g., 10-15%), although I expressed my concern about this altogether [here](https://github.com/cowprotocol/services/issues/740) as well, labeling it as order censorship in a way.

**Review by nlordell — APPROVED** [review]
LGTM

**Pull request merged**